src/IW/Sync/Search.hs

```diff
@@ -28,70 +35,117 @@ import qualified Data.Text as T
 import qualified Data.Vector as V
 
 
--- | Fetch all repositories built with the Haskell language.
-fetchAllHaskellRepos
-    :: ( MonadIO m
+githubSearch
+    :: forall a env m.
+       ( MonadIO m
        , WithError m
        , WithLog env m
+       , FromJSON a
+       , Typeable a
        )
-    => m [GitHub.Repo]
-fetchAllHaskellRepos = liftGitHubSearchToApp searchHaskellRepos
+    => [Text] -- ^ Paths
+    -> Text   -- ^ Query String
+    -> Day    -- ^ From day
+    -> Day    -- ^ To day
+    -> Int    -- ^ Page
+    -> m [a]
+githubSearch paths queryString from to page = liftIO (executeRequest' $ query paths queries) >>= \case
+    Left err -> throwError $ githubErrToAppErr err
+    Right (SearchResult count vec) -> do
+        log Info $ "Query executed with the following paths and query parameters: " <> show paths <> " " <> show queries
+        log Info $ "Fetched total of " <> show count <> " " <> typeName @a <> "s..."
+        searchLimit <- getSearchRateLimit
+        log Info $ "Search rate limit information: " <> show searchLimit
+        pure $ V.toList vec
+  where
+    queries :: [(ByteString, Maybe ByteString)]
+    queries = pagination <> [("q", Just $ encodeUtf8 $ strip queryString <> " " <> dateRange from to)]
 
--- | Convert a value of the @GitHub.Repo@ type to a value of our own @Repo@ type.
-fromGitHubRepo :: GitHub.Repo -> Repo
-fromGitHubRepo githubRepo = Repo
-    { repoOwner      = RepoOwner $ GitHub.untagName $ GitHub.simpleOwnerLogin $ GitHub.repoOwner githubRepo
-    , repoName       = RepoName  $ GitHub.untagName $ GitHub.repoName githubRepo
-    , repoDescr      = fromMaybe "" $ GitHub.repoDescription githubRepo
-    , repoCategories = SqlArray []
-    }
+    pagination :: [(ByteString, Maybe ByteString)]
+    pagination =
+        [ ("per_page", Just "100")
+        , ("page", Just $ show page)
+        ]
+
+    dateRange :: Day -> Day -> Text
+    dateRange from' to' = "created:" <> julianDayToIso from' <> ".." <> julianDayToIso to'
+
+    julianDayToIso :: Day -> Text
+    julianDayToIso = fromString . formatTime defaultTimeLocale (iso8601DateFormat Nothing)
+
+getSearchRateLimit
+    :: forall m.
+       ( MonadIO m
+       , WithError m
+       )
+    => m Limits
+getSearchRateLimit = liftIO rateLimit >>= \case
+    Left err -> throwError $ githubErrToAppErr err
+    Right RateLimit{..} -> pure rateLimitSearch
+
+-- | Fetch all repositories built with the Haskell language by page within a date range.
+fetchHaskellReposByDate
+    :: forall env m.
+       ( MonadIO m
+       , WithError m
+       , WithLog env m
+       )
+    => Day
+    -> Day
+    -> Int
+    -> m [GitHub.Repo]
+fetchHaskellReposByDate from to page = githubSearch paths queryString from to page
```

Comment:
Can you elaborate on the meaning of `paths` argument? I don't quite understand why it's need and why it's a list of `Text`... Also, it seems like this function can be written shorter and there's no need to introduce variables in `where` clause. Also can eta-reduce the function. Something like:

```haskell
fetchHaskellReposByDate = githubSearch ["search", "repositories"] "language:haskell"
```

Reply:
The paths arguement is for this function: 
```
query paths queries
```
It constructs a github query given a list of paths and list of query parameter/value pairs. I agree that this function can be a lot simpler. I will fix this.

Reply:
I forgot to use `Paths` and `QueryString` type synonyms from `github` library. If you look at the new code it should make sense. Maybe still, using `Paths` might be overkill but I prefer it over just a `Text`.